gcc/cmassoc/offset.1
.TH offset 1 "August 2013" "cmassoc-tools-1.9.1" "Motley Tools"

.SH NAME
offset - data offset table generator

.SH SYNOPSIS
.BR offset
.IR file

.SH DESCRIPTION
Read an object definition file and print a data offset table on stdout in various forms.
The default output is in plain text that can be cut-and-pasted into email messages and technical documents.

.PP
A data offset table lists data objects in order showing their offset, length, symbol and description.
This information is useful to programmers working with large or complex data structures.

.SH OPTIONS

.TP
.RB - b
Print a Docbook table on stdout.
Object groups are displayed as separate tables separated by headers.
The output can be included in the body of an existing HTML page.

.TP
-\fBc \fIcolumn\fR
Align object descriptions to this column, if possible, for text output.
If this option is omitted then description will immediately follow the object names.
This option has no effect for HTML output.

.TP
.RB - e
Print in efsu format with comments.
This option can be used to communicate an Ethernet frame layout to someone with access to program \fBefsu\fR.

.TP
.RB - h
Print an HTML table on stdout.
Object groups are displayed as separate tables separated by headers.
The output can be included in the body of an existing HTML page.

.TP
.RB - H
Print an HTML webpage on stdout.
The page consists of an HTML head and body.
The body contains the same output as that generated by option \fB-t\fR.

.TP
.RB - r
Reset at each heading.
This option only has effect when used with option \fB-x\fR.
The grouping of unmarked objects resets at each new heading.
This ensures that headings are preserved and displayed.

.TP
.RB - s
Print a CSS stylesheet on stdout where the classes correspond to those used in HTML formatted output.
This output and be included in a existing CSS stylesheet or used to create a new CSS stylesheet.

.TP
.RB - t
Print as plain text table with descriptions aligned as defined by option \fB-c\fR.

.TP
.RB - x
Fold unmarked objects into single values.
Marked objects have a '\fB+\fR' or '\fB-\fR' prefix.
Unmarked objects, are combined into one "\FBRSVD\fR" object which effecively hides the offset, length and name of the individual objects.

.TP
.RB - z
Remove leading zeros on object length values.
Leading zeros determine the background color of table entries and are used to highlight selected entries.
This option effectively removes all background colors for all objects.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over previous command line options.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over previous command line options.

.SH ARGUMENTS

.TP
\fIfile\fR
The name of an object definition file.
An object definition file entry has three fields: \fBsize\fR in bytes, \fBname\fR and \fBdescription\fR.
Size is expressed as an unsigned decimal integer.
Name consists of consecutive letters, digits, hyphens, colons and underscores.
Description consists of any remaining text to the end of the line.
Objects with a size of 0 bytes are treated as a heading.

.SH EXAMPLES
The following example reads offset definition file \fBobjects.txt\fR and outputs text showing object offsets, lengths, names and descriptions.
The offset and size of object groups is computed from those of group members and displayed, as well.

.PP
   # offset objects.txt

.SH SEE ALSO
.BR efsu ( 1 )
.BR html ( 1 ),
.BR matrix ( 1 )

.SH CREDITS
 Charles Maier <cmaier@cmassoc.net>



gcc/cmassoc/offset.c
/*====================================================================*
 *
 *   offset.c - print offset table;  
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <ctype.h>
#include <errno.h>
#include <limits.h>
#include <string.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/cmassoc.h"
#include "../tools/markup.h"
#include "../tools/format.h"
#include "../linux/linux.h"
#include "../date/date.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/console.c"
#include "../tools/efreopen.c"
#include "../tools/emalloc.c"
#include "../tools/uintspec.c"
#include "../tools/todigit.c"
#include "../tools/indent.c"
#endif

#ifndef MAKEFILE
#include "../linux/hostname.c"
#include "../linux/username.c"
#include "../linux/getusername.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define OFFSET_HTML (1 << 0)
#define OFFSET_PAGE (1 << 1)
#define OFFSET_TEXT (1 << 2)
#define OFFSET_EFSU (1 << 3)
#define OFFSET_BOOK (1 << 4)
#define OFFSET_FOLD (1 << 5)
#define OFFSET_HOLE (1 << 6)
#define OFFSET_ZERO (1 << 7)

#define COLUMN 0
#define MARGIN 0

/*====================================================================*
 *   program variables;
 *--------------------------------------------------------------------*/

static char const * style = "offset"; 
static char const * style_posted = "posted"; 
static char const * style_offset = "offset"; 
static char const * style_length = "length"; 
static char const * style_symbol = "symbol"; 
static char const * style_string = "string"; 
static char const * label_offset = "offset"; 
static char const * label_length = "length"; 
static char const * label_symbol = "symbol"; 
static char const * label_string = "description"; 
static unsigned margin = MARGIN; 
static unsigned column = COLUMN; 
static unsigned lineno = 0; 
static unsigned origin = 0; 
static unsigned offset = 0; 
static unsigned length = 0; 
static char * symbol = (char *)(0); 
static char * string = (char *)(0); 
static signed c; 

/*====================================================================*
 *   
 *   unsigned object ();
 *
 *   scan and parse an object definition;
 *
 *   <definition> := <offset> <symbol> <string>
 *
 *   <symbol> := <alpha> | '_'
 *   <symbol> := <symbol> <alpha> 
 *   <symbol> := <symbol> '-' 
 *   <symbol> := <symbol> '_' 
 *   <symbol> := <symbol> '.' 
 *   <symbol> := <symbol> ':' 
 *   <symbol> := <symbol> '[' <number> ']'
 *
 *   <string> := <alpha> | <digit> | <punct>
 *   <string> := <string> <alpha>
 *   <string> := <string> <digit>
 *   <string> := <string> <punct>
 *   <string> := <string> <blank>
 *   
 *   <number> := <digit>
 *   <number> := <number> <digit>
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned object () 

{ 
	extern signed c; 
	extern unsigned lineno; 
	extern unsigned length; 
	extern char * symbol; 
	extern char * string; 
	unsigned count = 0; 
	char * sp; 
	char * cp; 
	length = 0; 
	if ((c == '+') || (c == '-')) 
	{ 
		do 
		{ 
			c = getc (stdin); 
		} 
		while (isblank (c)); 
	} 
	while (c == '0') 
	{ 
		c = getc (stdin); 
		count++; 
	} 
	while (isdigit (c)) 
	{ 
		length *= 10; 
		length += c - '0'; 
		c = getc (stdin); 
	} 
	while (isblank (c)) 
	{ 
		c = getc (stdin); 
	} 
	sp = symbol; 
	if (isalpha (c) || (c == '_')) 
	{ 
		do 
		{ 
			* sp++ = (char)(c); 
			c = getc (stdin); 
		} 
		while (isalnum (c) || (c == '_') || (c == '-') || (c == '.') || (c == ':')); 
	} 
	while (isblank (c)) 
	{ 
		c = getc (stdin); 
	} 
	if (c == '[') 
	{ 
		* sp++ = (char)(c); 
		c = getc (stdin); 
		while (isblank (c)) 
		{ 
			c = getc (stdin); 
		} 
		while (isdigit (c)) 
		{ 
			* sp++ = (char)(c); 
			c = getc (stdin); 
		} 
		while (isblank (c)) 
		{ 
			c = getc (stdin); 
		} 
		* sp = (char)(0); 
		if (c != ']') 
		{ 
			error (1, EINVAL, "Have '%s' but need ']' on line %d", symbol, lineno); 
		} 
		* sp++ = (char)(c); 
		c = getc (stdin); 
	} 
	* sp = (char)(0); 
	while (isblank (c)) 
	{ 
		c = getc (stdin); 
	} 
	sp = cp = string; 
	while (nobreak (c)) 
	{ 
		* cp++ = (char)(c); 
		if (!isspace (c)) 
		{ 
			sp = cp; 
		} 
		c = getc (stdin); 
	} 
	* sp = (char)(0); 
	return (count); 
} 

/*====================================================================*
 *
 *   unsigned stylesheet (unsigned margin);
 *   
 *   print css stylesheet on stdout; this ensures that element class 
 *   names agree;
 *   
 *   table.offset { border: solid 1pt black; border-spacing: 1px; background: white; } 
 *   th.offset, td.offset { border: solid 1pt silver; text-align: right; width: 080px; }
 *   th.length, td.length { border: solid 1pt silver; text-align: right; width: 080px; }
 *   th.symbol, td.symbol { border: solid 1pt silver; text-align: right; width: 300px; }
 *   th.string, td.string { border: solid 1pt silver; }
 *   
 *   table.data { border: solid 1pt black; border-spacing: 1px; background: white; } 
 *   th.base, td.base { border: solid 1pt silver; text-align: right; width: 080px; }
 *   th.size, td.size { border: solid 1pt silver; text-align: right; width: 080px; }
 *   th.name, td.name { border: solid 1pt silver; text-align: right; width: 300px; }
 *   th.text, td.text { border: solid 1pt silver; }
 *   div.posted { text-align: center; margin: 10px; }
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned stylesheet (unsigned margin) 

{ 
	indent (margin, "table { table-layout: fixed; background: transparent; border: solid 1pt black; border-collapse: separate; border-spacing: 1px; font: normal 10pt verdana; }"); 
	indent (margin, "th { background: inherit; border: solid 1pt silver; padding: 2px 10px; text-align: center; vertical-align: middle; }"); 
	indent (margin, "td { background: inherit; border: solid 1pt silver; padding: 2px 10px; text-align: left; vertical-align: top; }"); 
	indent (margin, "th.%s { width: 080px; }", style_offset); 
	indent (margin, "th.%s { width: 080px; }", style_length); 
	indent (margin, "th.%s { width: 300px; }", style_symbol); 
	indent (margin, "th.%s { }", style_string); 
	indent (margin, "td.%s { text-align: right; }", style_offset); 
	indent (margin, "td.%s { text-align: right; }", style_length); 
	indent (margin, "td.%s { text-align: right; }", style_symbol); 
	indent (margin, "td.%s { text-align: left;  }", style_string); 
	indent (margin, "div.%s { text-align: center; margin: 10px; };", style_posted); 
	return (margin); 
} 

/*====================================================================*
 *   
 *   void html (char const * colors [], unsigned count, flag_t flags);
 *
 *   print object definitions in HTML format with headings and color
 *   coded backgrounds;
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void html (char const * colors [], unsigned count, flag_t flags) 

{ 
	time_t now = time (& now); 
	static char datetime [LOGTIME_LEN]; 
	extern unsigned lineno; 
	extern unsigned margin; 
	extern unsigned offset; 
	extern unsigned length; 
	extern char * symbol; 
	extern char * string; 
	extern signed c; 
	strftime (datetime, sizeof (datetime), LOGTIME, localtime (& now)); 
	lineno = 1; 
	offset = 0; 
	length = 0; 
	if (_anyset (flags, OFFSET_PAGE)) 
	{ 
		indent (margin++, "<html>"); 
		indent (margin++, "<title>"); 
		indent (margin--, "</title>"); 
		indent (margin++, "<head>"); 
		indent (margin++, "<style>"); 
		margin = stylesheet (margin); 
		indent (margin--, "</style>"); 
		indent (margin--, "</head>"); 
		indent (margin++, "<body>"); 
		indent (0, "<!-- BEGIN CONTENT -->"); 
	} 
	while ((c = getc (stdin)) != EOF) 
	{ 
		unsigned index; 
		if (isspace (c)) 
		{ 
			if (c == '\n') 
			{ 
				lineno++; 
			} 
			continue; 
		} 
		if ((c == '#') || (c == ';')) 
		{ 
			do 
			{ 
				c = getc (stdin); 
			} 
			while (nobreak (c)); 
			lineno++; 
			continue; 
		} 
		index = object () % count; 
		if (!length) 
		{ 
			if (offset) 
			{ 
				indent (margin--, "</table>"); 
			} 
			indent (margin++, "<h2 class='%s'>", style); 
			indent (margin, "%s %s", symbol, string); 
			indent (margin--, "</h2>"); 
			indent (margin++, "<table class='%s'>", style); 
			indent (margin++, "<tr class='%s'>", style); 
			indent (margin++, "<th class='%s'>", style_offset); 
			indent (margin, "%s", label_offset); 
			indent (margin--, "</th>"); 
			indent (margin++, "<th class='%s'>", style_length); 
			indent (margin, "%s", label_length); 
			indent (margin--, "</th>"); 
			indent (margin++, "<th class='%s'>", style_symbol); 
			indent (margin, "%s", label_symbol); 
			indent (margin--, "</th>"); 
			indent (margin++, "<th class='%s'>", style_string); 
			indent (margin, "%s", label_string); 
			indent (margin--, "</th>"); 
			indent (margin--, "</tr>"); 
		} 
		else 
		{ 
			if (!offset) 
			{ 
				indent (margin++, "<h2 class='%s'>", style); 
				indent (margin, "No Title"); 
				indent (margin--, "</h2>"); 
				indent (margin++, "<table>"); 
				indent (margin++, "<table class='%s'>", style); 
				indent (margin++, "<tr class='%s'>", style); 
				indent (margin++, "<th class='%s'>", style_offset); 
				indent (margin, "%s", label_offset); 
				indent (margin--, "</th>"); 
				indent (margin++, "<th class='%s'>", style_length); 
				indent (margin, "%s", label_length); 
				indent (margin--, "</th>"); 
				indent (margin++, "<th class='%s'>", style_symbol); 
				indent (margin, "%s", label_symbol); 
				indent (margin--, "</th>"); 
				indent (margin++, "<th class='%s'>", style_string); 
				indent (margin, "%s", label_string); 
				indent (margin--, "</th>"); 
				indent (margin--, "</tr>"); 
			} 
			indent (margin++, "<tr class='%s'>", style); 
			indent (margin++, "<td class='%s' style='background: %s;'>", style_offset, colors [index]); 
			indent (margin, "%08X", offset); 
			indent (margin--, "</td>"); 
			indent (margin++, "<td class='%s' style='background: %s;'>", style_length, colors [index]); 
			indent (margin, "%6d", length); 
			indent (margin--, "</td>"); 
			indent (margin++, "<td class='%s' style='background: %s;'>", style_symbol, colors [index]); 
			indent (margin, "%s", * symbol? symbol: "&nbsp;"); 
			indent (margin--, "</td>"); 
			indent (margin++, "<td class='%s' style='background: %s;'>", style_string, colors [index]); 
			indent (margin, "%s", * string? string: "&nbsp;"); 
			indent (margin--, "</td>"); 
			indent (margin--, "</tr>"); 
		} 
		offset += length; 
		lineno++; 
	} 
	if (offset) 
	{ 
		indent (margin--, "</table>"); 
	} 
	indent (margin++, "<div class='%s'>", style_posted); 
	indent (margin, "Posted %s on %s by %s", datetime, hostname (), username (getuid ())); 
	indent (margin--, "</div>"); 
	indent (0, "<!-- END CONTENT -->"); 
	if (_anyset (flags, OFFSET_PAGE)) 
	{ 
		indent (margin--, "</body>"); 
		indent (margin--, "</html>"); 
	} 
	return; 
} 

/*====================================================================*
 *   
 *   void text (flag_t flags);
 *   
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void text (flag_t flags) 

{ 
	extern unsigned lineno; 
	extern unsigned margin; 
	extern unsigned column; 
	extern unsigned origin; 
	extern unsigned offset; 
	extern unsigned length; 
	extern char * symbol; 
	extern char * string; 
	extern signed c; 
	lineno = 1; 
	origin = 0; 
	offset = 0; 
	length = 0; 
	while ((c = getc (stdin)) != EOF) 
	{ 
		if (isspace (c)) 
		{ 
			if (c == '\n') 
			{ 
				lineno++; 
			} 
			continue; 
		} 
		if ((c == '#') || (c == ';')) 
		{ 
			do 
			{ 
				c = getc (stdin); 
			} 
			while (nobreak (c)); 
			lineno++; 
			continue; 
		} 
		object (); 
		if (length) 
		{ 
			unsigned output = printf (" %08X %10d %s", offset, length, symbol); 
			while (output++ < column) 
			{ 
				putc (' ', stdout); 
			} 
			if (* string) 
			{ 
				printf (" (%s)", string); 
			} 
			printf ("\n"); 
		} 
		else 
		{ 
			if (offset) 
			{ 
				printf (" -------- %10d bytes\n\n", offset - origin); 
			} 
			printf (" -------- %10d %s %s\n", offset, symbol, string); 
			origin = offset; 
		} 
		offset += length; 
		lineno++; 
	} 
	if (offset) 
	{ 
		printf (" -------- %10d bytes\n\n", offset - origin); 
	} 
	printf (" %08X  %10d bytes\n", offset, offset); 
	return; 
} 

/*====================================================================*
 *   
 *   void efsu (flag_t flags);
 *   
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void efsu (flag_t flags) 

{ 
	extern unsigned lineno; 
	extern unsigned margin; 
	extern unsigned column; 
	extern unsigned origin; 
	extern unsigned offset; 
	extern unsigned length; 
	extern char * symbol; 
	extern char * string; 
	extern signed c; 
	lineno = 1; 
	origin = 0; 
	offset = 0; 
	length = 0; 
	while ((c = getc (stdin)) != EOF) 
	{ 
		if (isspace (c)) 
		{ 
			if (c == '\n') 
			{ 
				lineno++; 
			} 
			continue; 
		} 
		if ((c == '#') || (c == ';')) 
		{ 
			do 
			{ 
				c = getc (stdin); 
			} 
			while (nobreak (c)); 
			lineno++; 
			continue; 
		} 
		object (); 
		if (length) 
		{ 
			unsigned column = 0; 
			printf ("# %s\n", symbol); 
			while (column < length) 
			{ 
				putc ('0', stdout); 
				putc ('0', stdout); 
				putc (++ column % 16? ' ': '\n', stdout); 
			} 
			if (column % 16) 
			{ 
				putc ('\n', stdout); 
			} 
		} 
		offset += length; 
		lineno++; 
	} 
	return; 
} 

/*====================================================================*
 *   
 *   void tabs (flag_t flags);
 *   
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void tabs (flag_t flags) 

{ 
	extern unsigned lineno; 
	extern unsigned margin; 
	extern unsigned offset; 
	extern unsigned length; 
	extern char * symbol; 
	extern char * string; 
	extern signed c; 
	lineno = 1; 
	offset = 0; 
	length = 0; 
	while ((c = getc (stdin)) != EOF) 
	{ 
		if (isspace (c)) 
		{ 
			if (c == '\n') 
			{ 
				lineno++; 
			} 
			continue; 
		} 
		if ((c == '#') || (c == ';')) 
		{ 
			do 
			{ 
				c = getc (stdin); 
			} 
			while (nobreak (c)); 
			lineno++; 
			continue; 
		} 
		object (); 
		if (length) 
		{ 
			printf ("%08X\t%6d\t%s\t%s\n", offset, length, symbol, string); 
		} 
		else 
		{ 
			printf ("\n\t\t%s %s\n", symbol, string); 
		} 
		offset += length; 
		lineno++; 
	} 
	return; 
} 

/*====================================================================*
 *   
 *   void fold (flag_t flags);
 *   
 *   fold unmarked objects into single reserved objects;
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void fold (flag_t flags) 

{ 
	unsigned offset = 0; 
	unsigned extent = 0; 
	unsigned length = 0; 
	signed c = getc (stdin); 
	while (c != EOF) 
	{ 
		while (isspace (c)) 
		{ 
			c = getc (stdin); 
		} 
		if (c == '+') 
		{ 
			do 
			{ 
				c = getc (stdin); 
			} 
			while (isblank (c)); 
			while (isdigit (c)) 
			{ 
				length *= 10; 
				length += c - '0'; 
				c = getc (stdin); 
			} 
			if (extent) 
			{ 
				printf ("0%d RSVD\n", extent); 
				extent = 0; 
			} 
			printf ("%4d", length); 
			while (nobreak (c)) 
			{ 
				putc (c, stdout); 
				c = getc (stdin); 
			} 
			putc ('\n', stdout); 
			offset += length; 
			length = 0; 
			continue; 
		} 
		if (isdigit (c)) 
		{ 
			while (isdigit (c)) 
			{ 
				length *= 10; 
				length += c - '0'; 
				c = getc (stdin); 
			} 
			if (!length) 
			{ 
				if (_anyset (flags, OFFSET_HOLE)) 
				{ 
					if (extent) 
					{ 
						printf ("0%d RSVD\n", extent); 
						extent = 0; 
					} 
					if (offset) 
					{ 
						putc ('\n', stdout); 
					} 
					printf ("%4d", length); 
					while (nobreak (c)) 
					{ 
						putc (c, stdout); 
						c = getc (stdin); 
					} 
					putc ('\n', stdout); 
					continue; 
				} 
			} 
			while (nobreak (c)) 
			{ 
				c = getc (stdin); 
			} 
			extent += length; 
			offset += length; 
			length = 0; 
			continue; 
		} 
		while (nobreak (c)) 
		{ 
			putc (c, stdout); 
			c = getc (stdin); 
		} 
	} 
	if (extent) 
	{ 
		printf ("0%d RSVD\n", extent); 
		extent = 0; 
	} 
	return; 
} 

/*====================================================================*
 *   
 *   void zero (flag_t flags);
 *   
 *   read object description file and remove leading 0's from length
 *   specifications to remove color coding;
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void zero (flag_t flags) 

{ 
	signed c = getc (stdin); 
	while (c != EOF) 
	{ 
		while (isspace (c)) 
		{ 
			putc (c, stdout); 
			c = getc (stdin); 
		} 
		if ((c == '+') || (c == '-')) 
		{ 
			do 
			{ 
				putc (c, stdout); 
				c = getc (stdin); 
			} 
			while (isblank (c)); 
		} 
		while (c == '0') 
		{ 
			c = getc (stdin); 
			if (!isdigit (c)) 
			{ 
				putc ('0', stdout); 
				break; 
			} 
		} 
		while (nobreak (c)) 
		{ 
			putc (c, stdout); 
			c = getc (stdin); 
		} 
	} 
	return; 
} 

/*====================================================================*
 *
 *   void function (char const * colors [], unsigned count, flag_t flags);
 *
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void function (char const * colors [], unsigned count, flag_t flags) 

{ 
	if (_anyset (flags, (OFFSET_HTML | OFFSET_PAGE))) 
	{ 
		html (colors, count, flags); 
	} 
	else if (_anyset (flags, (OFFSET_TEXT))) 
	{ 
		tabs (flags); 
	} 
	else if (_anyset (flags, (OFFSET_EFSU))) 
	{ 
		efsu (flags); 
	} 
	else if (_anyset (flags, (OFFSET_FOLD))) 
	{ 
		fold (flags); 
	} 
	else if (_anyset (flags, (OFFSET_ZERO))) 
	{ 
		zero (flags); 
	} 
	else 
	{ 
		text (flags); 
	} 
	return; 
} 

/*====================================================================*
 *   
 *   int main (int argc, char const * argv []);
 *   
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{ 
	extern unsigned margin; 
	extern unsigned column; 
	static char const * optv [] = 
	{ 
		"bc:ehl:prstxz", 
		PUTOPTV_S_FUNNEL, 
		"print offset table", 
		"b\tprint docbook format", 
		"c n\talign descriptions to column (n) [" LITERAL (COLUMN) "]", 
		"e\tprint efsu format", 
		"h\tprint HTML table on stdout", 
		"l n\tindent level is (n) [" LITERAL (MARGIN) "]", 
		"p\tprint HTML page on stdout", 
		"r\treset at headings", 
		"s\tprint CSS2 stylesheet on stdout", 
		"t\tprint text with TAB seperated columns", 
		"x\thide unmarked objects", 
		"z\tremove leading zeros", 
		(char const *)(0)
	}; 
	char const * colors [] = 
	{ 
		"#FFFFFF", 
		"#FFFF00", 
		"#00FFFF", 
		"#00FF00", 
		"#FF00FF"
	}; 
	flag_t flags = (flag_t)(0); 
	signed c; 
	while (~ (c = getoptv (argc, argv, optv))) 
	{ 
		switch (c) 
		{ 
		case 'b': 
			_setbits (flags, OFFSET_BOOK); 
			break; 
		case 'c': 
			column = uintspec (optarg, 0, UCHAR_MAX); 
			break; 
		case 'e': 
			_setbits (flags, OFFSET_EFSU); 
			break; 
		case 'h': 
			_setbits (flags, OFFSET_HTML); 
			break; 
		case 'p': 
			_setbits (flags, OFFSET_PAGE); 
			break; 
		case 'l': 
			margin = (unsigned) (uintspec (optarg, 0, 16)); 
			break; 
		case 'r': 
			_setbits (flags, OFFSET_HOLE); 
			break; 
		case 's': 
			stylesheet (margin); 
			return (0); 
		case 't': 
			_setbits (flags, OFFSET_TEXT); 
			break; 
		case 'x': 
			_setbits (flags, OFFSET_FOLD); 
			break; 
		case 'z': 
			_setbits (flags, OFFSET_ZERO); 
			break; 
		default: 
			break; 
		} 
	} 
	argc -= optind; 
	argv += optind; 
	symbol = emalloc (SYMBOLSIZE); 
	string = emalloc (STRINGSIZE); 
	if (!argc) 
	{ 
		function (colors, sizeof (colors) / sizeof (const char *), flags); 
	} 
	while ((argc) && (* argv)) 
	{ 
		if (efreopen (* argv, "rb", stdin)) 
		{ 
			function (colors, SIZEOF (colors), flags); 
		} 
		argc--; 
		argv++; 
	} 
	return (0); 
} 

